### ISSUE TITLE
Pyright thinks that int/str are not comparable

### ISSUE THREAD

**Issue by heatherleaf** [microsoft/pyright#5432]
**Describe the bug**
I tried to implement a type for comparables, according to this suggestion: https://stackoverflow.com/a/65224102 

**To Reproduce**
Run pyright on the code below.

**Expected behavior**
There should be no type errors.

**Code or Screenshots**

```python
from typing import Protocol, TypeVar, Any

class SupportsLessThan(Protocol):
    def __lt__(self: 'Comparable', other: 'Comparable') -> bool: ...
Comparable = TypeVar("Comparable", bound=SupportsLessThan)

def less_than(a: Comparable, b: Comparable) -> bool:
    return a < b

less_than(3, 4)  # error: Argument of type "Literal[3]" cannot be assigned to parameter "a" of type "Comparable@less_than"

def sorted_list(xs: list[Comparable]) -> list[Comparable]:
    return sorted(xs)

sorted_list(["b", "a"])  # error: Argument of type "list[str]" cannot be assigned to parameter "xs" of type "list[Comparable@sorted_list]"
```

Mypy reports no issues:
```
$ mypy tester.py 
Success: no issues found in 1 source file
```

But pyright does:
```
$ pyright tester.py 
/Users/peb/Documents/Undervisning/Datastrukturer/Repositories/labs/labs/binary-search/python/solution/tester.py
  /Users/peb/Documents/Undervisning/Datastrukturer/Repositories/labs/labs/binary-search/python/solution/tester.py:72:11 - error: Argument of type "Literal[3]" cannot be assigned to parameter "a" of type "Comparable@less_than" in function "less_than"
    Type "Literal[3]" cannot be assigned to type "SupportsLessThan"
      "Literal[3]" is incompatible with protocol "SupportsLessThan"
        "__lt__" is an incompatible type
          Type "(__value: int, /) -> bool" cannot be assigned to type "(other: Literal[3]) -> bool"
            Position-only parameter mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
  /Users/peb/Documents/Undervisning/Datastrukturer/Repositories/labs/labs/binary-search/python/solution/tester.py:72:14 - error: Argument of type "Literal[4]" cannot be assigned to parameter "b" of type "Comparable@less_than" in function "less_than"
    Type "Literal[4]" cannot be assigned to type "SupportsLessThan"
      "Literal[4]" is incompatible with protocol "SupportsLessThan"
        "__lt__" is an incompatible type
          Type "(__value: int, /) -> bool" cannot be assigned to type "(other: Literal[4]) -> bool"
            Position-only parameter mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
  /Users/peb/Documents/Undervisning/Datastrukturer/Repositories/labs/labs/binary-search/python/solution/tester.py:74:13 - error: Argument of type "list[str]" cannot be assigned to parameter "xs" of type "list[Comparable@sorted_list]" in function "sorted_list"
    Type "Literal['b']" cannot be assigned to type "SupportsLessThan"
      "Literal['b']" is incompatible with protocol "SupportsLessThan"
        "__lt__" is an incompatible type
          Type "(__value: str, /) -> bool" cannot be assigned to type "(other: Literal['b']) -> bool"
            Position-only parameter mismatch; expected 1 but received 0
    Type "Literal['a']" cannot be assigned to type "SupportsLessThan"
      "Literal['a']" is incompatible with protocol "SupportsLessThan"
        "__lt__" is an incompatible type (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations 
```

(The first two errors are for variables `a` and `b` in `less_than`, and the final error is for `sorted_list`)

Note: it does work for my own "comparable" classes (anything with an `__lt__` method), so it's only for built-in types I get this problem.

**VS Code extension or command-line**
Pyright v1.1.316
Pylance v2023.6.40

**Comment by heatherleaf** [comment]
I get exactly the same problem if I copy-paste the definitions from typeshed's `builtins.pyi`, and use them instead:

```python
_T_contra = TypeVar("_T_contra", contravariant=True)

class SupportsDunderLT(Protocol[_T_contra]):
    def __lt__(self, __other: _T_contra) -> bool: ...

class SupportsDunderGT(Protocol[_T_contra]):
    def __gt__(self, __other: _T_contra) -> bool: ...

SupportsRichComparison: TypeAlias = SupportsDunderLT[Any] | SupportsDunderGT[Any]
SupportsRichComparisonT = TypeVar("SupportsRichComparisonT", bound=SupportsRichComparison)  # noqa: Y001
```

**Comment by erictraut** [comment]
Your protocol definition indicates that the `__lt__` method must accept a named parameter called `other`, but the `__lt__` method in the `int` and `str` classes accept positional-only parameters. You can fix the problem in your code by adding a positional-only indicator in your protocol definition.

```python
class SupportsLessThan(Protocol):
    def __lt__(self: "Comparable", other: "Comparable", /) -> bool:
        ...
```

Alternatively, you can use the older mechanism for specifying a positional-only parameter by naming it with a double underscore. This is the technique used in typeshed.

```python
class SupportsLessThan(Protocol):
    def __lt__(self: "Comparable", __other: "Comparable") -> bool:
        ...
```

Once you fix this issue in your code, it type checks without error.

Mypy does not enforce positional-only parameters in protocols, but it should. If I remember correctly, there's an open bug in the mypy issue tracker.

I'm going to close this issue because pyright is doing the right thing here. This isn't a bug.
